### ISSUE TITLE
Suggestion: if bookmark is an image, use the image as the thumbnail

### ISSUE THREAD

**Issue by cengels** [perfect-things/perfect-home#77]
Hi! First of all, I am liking the extension quite a bit. It's nice, simple, and clean. As for the suggestion:

I am using PerfectHome as an image gallery of sorts. That is, I bookmark local images and arrange them in folders as a sort of "album." As such, it is vital that all the images have appropriate thumbnails. Now, I understand it is not easy to get a thumbnail for a full website—but if the bookmark already leads to an image, all that needs to be done is that the thumbnail uses the image.

In fact, you can already do this by manually setting the bookmark thumbnail to lead to the same local (or non-local) file link. It's basically just a copy-and-paste; you don't need to modify the link in any way. However, doing this for hundreds of images is a lot of work. So my proposal is that, if the bookmark leads to an image, the image should automatically be used as the thumbnail _**if and only if**_ the user did not define a custom thumbnail for the bookmark.

Let me know what you think. Personally I would really love this feature.

**Comment by tborychowski** [comment]
Hi,
I'm happy that you like it.

>I am using PerfectHome as an image gallery of sorts

Ha! I'd never foreseen this use case! :-)
This makes complete sense, and definitely needs to be fixed!
I've just never had a bookmark to an image.

**Comment by tborychowski** [comment]
Sorry, I just tried that and unfortunately it's impossible to do.
Firefox security restrictions prevent extensions from accessing local files.

**Comment by cengels** [comment]
Wait, why is it impossible? With the workaround outlined in [the FAQ](https://github.com/perfect-things/perfect-home/blob/master/FAQ.md#cant-open-a-bookmark-to-a-local-file-starting-with-file) and a simple `Image()` check with `onload` and `onerror`  this kind of thing has worked for me before on local HTML sites. Do extensions behave differently in this regard?

**Comment by tborychowski** [comment]
Yes, even with the workaround it seems that extensions are more restricted.
I've just did a quick test: created a simple html file with 2 images included, e.g.:
```html
<image src="file:///Users/tom/screenshot.png">
<div style="width: 400px; height: 400px; background: url('file:///Users/tom/screenshot.png');"></div>
```
It works if I open that html file directly in a browser.
If I add these 2 lines to the extension - it doesn't!

The image `onerror` throws an error (without any messages):
<img width="583" alt="Screenshot 2020-04-09 at 11 09 38" src="https://user-images.githubusercontent.com/392513/78884381-b91fd700-7a52-11ea-8b42-3c93cc84382b.png">

**Comment by cengels** [comment]
Hm... Maybe I'll try a few different things myself when I have the time.

Until then, all is not necessarily lost—how about simply doing a string check for the file extension? If it's a known image extension (png, gif, bmp, jpg, jpeg, ...), set it as the thumbnail. Not a perfect solution for sure, but if I'm not mistaken, the worst thing that can happen is for the image to just not work. XSS, after all, isn't possible via the `background-image` property.

What do you think?

**Comment by tborychowski** [comment]
Let me rephrase my comment as it seems you did not understand it:

Local images do not work. 
No matter how you include them. 

> Maybe I'll try a few different things myself when I have the time.

good luck 😉

**Comment by cengels** [comment]
Well, that's clearly not true. I can explicitly declare a local image as the thumbnail to one of my bookmarks within Perfect Home and it works just fine. Maybe we're not talking about the same thing?

**Comment by tborychowski** [comment]
Can you show me a screenshot of the edit popup on a bookmark where you have that image "declared"?

There are 2 ways you can assign an image thumbnail for a bookmark:
- url to an image - where the (external) url is loaded as a thumbnail background image
- "browse" for an image from your computer, where the image is converted in a browser into a dataURL, and stored as such.

The second one works, because you select a file from your computer and you allow the browser to read this file.
The first one does not work (inside an extension).

**Comment by cengels** [comment]
Sure: 
![image](https://user-images.githubusercontent.com/31248098/78928397-d4193800-7aa0-11ea-8712-e81611a71d96.png)

As you can see, I did not use the *Browse* dialog, I just copy-pasted the file:/// url from the URL field to the thumbnail field and it worked without any extra work required. It's quite odd that it doesn't seem to work for your configuration.

**Comment by tborychowski** [comment]
You're right. I have no idea why this doesn't work on my machine 😕 
Maybe it's a mac issue?
I'll do some more testing.

**Comment by tborychowski** [comment]
I added a potential fix - javascript will try to load the image and if it succeeds - it will replace the URL of the thumbnail with that image. I cannot test this as I'm getting an error, so - when I release the new version - I'd ask you to do it, if you don't mind :-)

**Comment by cengels** [comment]
Of course! I'd be glad to. I'm sorry you couldn't find out why it doesn't work on your machine. If I notice any issues, I could take a look myself and open a pull request since I'd be able to directly test it, if you'd like.

**Comment by tborychowski** [comment]
OK, that's released now in v0.9.1.

**Comment by cengels** [comment]
Thank you so much! I did some preliminary testing and it seems to work very well for both file:/// URLs and non-file URLs. Explicitly setting a different thumbnail also works fine.

I did notice three things:

* When you open the edit popup, the image in the edit popup will always be the default placeholder, regardless of whether the bookmark leads to an image. I can see how this might be intended though to emphasize the fact that the thumbnail is "automatic."
* In the commit, I can see you're checking whether the URL protocol is "file" when you check whether it's an image. I may be reading this wrong, but if you're explicitly checking for "file," why does it work for non-file URLs as well (not that that's a bad thing!)?
* Some images don't work. Specifically when the file extension is not a common image file extension. For local files this is not a problem, but this means that some online images (like base64-encoded images or Twitter images) don't work. Personally I have no problem with this since I only bookmark local files anyway, but someone else might wonder.

Like I said, none of that bothers me in any way. I'm just noting it down because it might bother someone else. As far as I'm concerned though, you added exactly what I wanted, so thank you again.

**Comment by tborychowski** [comment]
Hi, Thanks for testing.

1. This was not intentional. The thumb should show up in edit popup. I'll fix that.
2. Yeah, I did that in the first commit, but I changed that in another one. I thought that if someone bookmarks image - they should see it anyway. The code in the first commit would cause the image to load twice (albeit from cache the second time) and - in case of some slowness in the system/lots of images - there would be a delay before thumbs show up. I hope that this not working on my machine is an outlier/edge case. If it doesn't work for someone, hopefully they will report it and I can improve then :-)
3. Yeah, I added the most common image extensions (https://github.com/perfect-things/perfect-home/blob/master/src/lib/utils.js#L194). If the URL doesn't end with image name - I have no way of testing it. The only way would be to try to load every url (for every bookmark) as an image in js (`const img = new Image(); img.src=...`) and check `onerror` callback, but that would slow down loading bookmarks a lot, and I don't want to do that.
